src/main/java/com/mentorship/vineservice/services/imp/OrderServiceImpl.java

```diff
@@ -0,0 +1,140 @@
+import com.mentorship.vineservice.services.OrderService;
+import com.mentorship.vineservice.services.VineService;
+import jakarta.transaction.Transactional;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+
+    private final OrderRepository orderRepository;
+    private final PostOfficeRepository postOfficeRepository;
+    private final VineMapper vineMapper;
+    private final VineService vineService;
+    private final RabbitTemplate rabbitTemplate;
+
+    private static final String CREATE_ORDER_EVENT_MESSAGE = "New order has been created";
+
+    @Value("${spring.rabbitmq.template.exchange}")
+    private String exchange;
+
+    @Value("${spring.rabbitmq.template.routing-key}")
+    private String routingKey;
+
+    @Override
+    public Long createOrder(OrderDto orderDto) throws VinePermissionException {
+
+        Order order = vineMapper.orderDtoToOrder(orderDto);
+        List<OrderVine> orderVineList = vineMapper.orderVineDtoToOrderVine(orderDto.getVines());
+
+        String homeAddress = orderDto.getDeliveryDetails().getHomeAddress();
+        PostOfficeDto postOffice = orderDto.getDeliveryDetails().getPostOffice();
+
+        validateDuplicationDeliveryDetails(homeAddress, postOffice);
+
+        addPostOfficeIfPresent(postOffice, order);
+
+        addVinesToOrder(order, orderVineList);
+
+        Long createdOrderId = orderRepository.save(order).getId();
+        log.info(String.format("New order created with id '%s", createdOrderId));
+
+        updateSoldVineAndVineAmount(orderVineList);
+
+        publishCreatedOrderEvent(CREATE_ORDER_EVENT_MESSAGE, createdOrderId);
+
+        return createdOrderId;
+    }
+
+
+    private void validateDuplicationDeliveryDetails(String homeAddress, PostOfficeDto postOffice)
+        throws VinePermissionException {
+        if (homeAddress == null && postOffice == null) {
+            throw new VinePermissionException(HttpStatus.BAD_REQUEST,
+                "'homeAddress' and 'postOffice' can not be null simultaneously");
+        }
+
+        if (homeAddress != null && postOffice != null) {
+            throw new VinePermissionException(HttpStatus.BAD_REQUEST,
+                "'homeAddress' and 'postOffice' can not be filled at the same order");
+        }
+    }
+
+    public void addPostOfficeIfPresent(PostOfficeDto postOffice, Order order) {
+        if (postOffice != null) {
+            PostOffice postOfficeObject = postOfficeRepository.findPostOfficeByCityAndOfficeNumber(postOffice.getCity(),
```

Comment:
It seems to me that this part can be moved to a separate service and then this class won't work with PostOfficeRepository, but will work with a separate service instead

Reply:
Yes, you are right 
Created PostOffficeService